Moonscraper Chart Editor/Assets/Scripts/Game/Charts/IO/Midi/MidReader.cs

```diff
@@ -340,21 +340,19 @@ private static void ReadNotes(IList<MidiEvent> track, Song song, Song.Instrument
             var noteProcessDict = GetNoteProcessDict(gameMode);
 
             if (instrument == Song.Instrument.Unrecognised)
+            {
+                var text = track[0] as TextEvent;
+                if (text != null)
```

Comment:
Check that "track" actually has events, might be dealing with an empty array and this will cause an index out of range exception in that case.

Reply:
That should be considered an invalid MIDI in the first place, and nowhere else in the code is this check made so it would cause an exception even before this point. It also wouldn't reach this point in the first place since that first event *has* to be a track event in order for this method to get called.

Reply:
If you still want it to be checked though, I can go ahead and do that at the first point it's relevant at.

Reply:
Yeah nah, because this stuff is local to a specific function then safety checks need to happen within the function scope. You shouldn't rely on the caller to have already done the safety checks in case this function ever gets reused or moved around incorrectly. 
That being said, the check before this point you pointed out should have the check as well. Both of this points should be logging an error when these cases come up as well.

Reply:
`ReadSync(midi.Events[0], song);` should also be changed to 
```
if (midi.Events.Count <= 0)
{
    throw new InvalidOperationException("Midi file has no tempo events, unable to parse.");
}
ReadSync(midi.Events[0], song);
```

Reply:
Got those checks added.